File: Sources/Sighting/WhisperService.swift
import AppKit
import Combine

/// Transkribiert Videosegmente lokal mit whisper.cpp (whisper-cli aus Homebrew).
/// Ablauf: ffmpeg extrahiert den In/Out-Bereich als 16-kHz-WAV,
/// whisper-cli transkribiert nach JSON, die Segmente landen mit
/// klickbaren Timecodes im Notizfeld.
final class WhisperService: ObservableObject {
    @Published var isRunning = false
    @Published var progressText = "Transkribiere …"
    @Published var lastError: String?
    @Published var isDownloadingModel = false
    @Published var downloadProgress: Double = 0

    static let modelURL = URL(fileURLWithPath: NSHomeDirectory())
        .appendingPathComponent("Library/Application Support/Sighting/models/ggml-large-v3-turbo.bin")
    static let remoteModelURL = URL(string:
        "https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-large-v3-turbo.bin")!

    private var downloadObservation: NSKeyValueObservation?

    static func findExecutable(_ name: String) -> String? {
        let candidates = [
            "/opt/homebrew/bin/\(name)",
            "/usr/local/bin/\(name)",
            "/usr/bin/\(name)",
        ]
        return candidates.first { FileManager.default.isExecutableFile(atPath: $0) }
    }

    var whisperPath: String? { Self.findExecutable("whisper-cli") }
    var ffmpegPath: String? { Self.findExecutable("ffmpeg") }
    var modelExists: Bool { FileManager.default.fileExists(atPath: Self.modelURL.path) }

    /// Was fehlt noch, bevor transkribiert werden kann?
    var setupProblem: String? {
        if whisperPath == nil {
            return "whisper-cli fehlt. Installation im Terminal: brew install whisper-cpp"
        }
        if ffmpegPath == nil {
            return "ffmpeg fehlt. Installation im Terminal: brew install ffmpeg"
        }
        if !modelExists {
            return "Das Whisper-Modell (~1,6 GB) muss einmalig geladen werden."
        }
        return nil
    }

    // MARK: Modell-Download

    func downloadModel() {
        guard !isDownloadingModel else { return }
        isDownloadingModel = true
        downloadProgress = 0
        lastError = nil

        let task = URLSession.shared.downloadTask(with: Self.remoteModelURL) { [weak self] tmp, _, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.isDownloadingModel = false
                self.downloadObservation = nil
                if let error {
                    self.lastError = "Modell-Download fehlgeschlagen: \(error.localizedDescription)"
                    return
                }
                guard let tmp else { return }
                do {
                    try FileManager.default.createDirectory(
                        at: Self.modelURL.deletingLastPathComponent(),
                        withIntermediateDirectories: true)
                    if FileManager.default.fileExists(atPath: Self.modelURL.path) {
                        try FileManager.default.removeItem(at: Self.modelURL)
                    }
                    try FileManager.default.moveItem(at: tmp, to: Self.modelURL)
                    self.objectWillChange.send()
                } catch {
                    self.lastError = "Modell konnte nicht gespeichert werden: \(error.localizedDescription)"
                }
            }
        }
        downloadObservation = task.progress.observe(\.fractionCompleted) { [weak self] progress, _ in
            DispatchQueue.main.async { self?.downloadProgress = progress.fractionCompleted }
        }
        task.resume()
    }

    // MARK: Transkription

    struct Segment {
        let start: Double   // Sekunden, relativ zum Video
        let text: String
    }

    struct TranscribeError: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    /// Transkribiert [start, end] des Videos und liefert Segmente mit
    /// Video-absoluten Timecodes.
    func transcribe(videoURL: URL, start: Double, end: Double,
                    completion: @escaping (Result<[Segment], TranscribeError>) -> Void) {
        guard let whisper = whisperPath, let ffmpeg = ffmpegPath, modelExists else {
            completion(.failure(TranscribeError(message: setupProblem ?? "Setup unvollständig")))
            return
        }
        isRunning = true
        progressText = "Extrahiere Audio …"
        lastError = nil

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let finish: (Result<[Segment], TranscribeError>) -> Void = { result in
                DispatchQueue.main.async {
                    self?.isRunning = false
                    if case .failure(let error) = result { self?.lastError = error.message }
                    completion(result)
                }
            }

            let tmpDir = FileManager.default.temporaryDirectory
                .appendingPathComponent("sighting-\(UUID().uuidString)")
            try? FileManager.default.createDirectory(at: tmpDir, withIntermediateDirectories: true)
            defer { try? FileManager.default.removeItem(at: tmpDir) }
            let wav = tmpDir.appendingPathComponent("segment.wav")

            // 1. Audio-Segment extrahieren
            let ffmpegResult = Self.run(ffmpeg, [
                "-hide_banner", "-loglevel", "error", "-nostdin",
                "-ss", String(start), "-to", String(end),
                "-i", videoURL.path,
                "-vn", "-ac", "1", "-ar", "16000", "-c:a", "pcm_s16le",
                wav.path,
            ])
            guard ffmpegResult.status == 0 else {
                finish(.failure(TranscribeError(message: "ffmpeg-Fehler: \(ffmpegResult.stderr.suffix(400))")))
                return
            }

            // 2. Whisper laufen lassen
            DispatchQueue.main.async { self?.progressText = "Transkribiere … 0 %" }
            let outBase = tmpDir.appendingPathComponent("transcript").path
            let whisperResult = Self.run(whisper, [
                "-m", Self.modelURL.path,
                "-f", wav.path,
                "-l", "auto",
                "-oj", "-of", outBase,
                "--print-progress",
            ]) { line in
                if let range = line.range(of: "progress ="),
                   let percent = line[range.upperBound...]
                       .split(separator: "%").first
                       .flatMap({ Int($0.trimmingCharacters(in: .whitespaces)) }) {
                    DispatchQueue.main.async {
                        self?.progressText = "Transkribiere … \(percent) %"
                    }
                }
            }
            guard whisperResult.status == 0 else {
                finish(.failure(TranscribeError(message: "Whisper-Fehler: \(whisperResult.stderr.suffix(400))")))
                return
            }

            // 3. JSON parsen
            let jsonURL = URL(fileURLWithPath: outBase + ".json")
            guard let data = try? Data(contentsOf: jsonURL),
                  let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let entries = root["transcription"] as? [[String: Any]] else {
                finish(.failure(TranscribeError(message: "Whisper-Ausgabe konnte nicht gelesen werden.")))
                return
            }
            let segments: [Segment] = entries.compactMap { entry in
                guard let text = (entry["text"] as? String)?
                        .trimmingCharacters(in: .whitespaces),
                      !text.isEmpty,
                      let offsets = entry["offsets"] as? [String: Any],
                      let fromMS = (offsets["from"] as? NSNumber)?.doubleValue
                else { return nil }
                return Segment(start: start + fromMS / 1000.0, text: text)
            }
            finish(.success(segments))
        }
    }

    // MARK: Prozess-Helfer

    private struct RunResult {
        let status: Int32
        let stdout: String
        let stderr: String
    }

    private static func run(_ launchPath: String, _ arguments: [String],
                            onStderrLine: ((String) -> Void)? = nil) -> RunResult {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: launchPath)
        process.arguments = arguments

        let outPipe = Pipe()
        let errPipe = Pipe()
        process.standardOutput = outPipe
        process.standardError = errPipe

        var stderrData = Data()
        let stderrLock = NSLock()
        errPipe.fileHandleForReading.readabilityHandler = { handle in
            let chunk = handle.availableData
            guard !chunk.isEmpty else { return }
            stderrLock.lock()
            stderrData.append(chunk)
            stderrLock.unlock()
            if let onStderrLine, let text = String(data: chunk, encoding: .utf8) {
                text.split(separator: "\n").forEach { onStderrLine(String($0)) }
            }
        }

        do {
            try process.run()
        } catch {
            return RunResult(status: -1, stdout: "", stderr: error.localizedDescription)
        }
        let stdoutData = outPipe.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()
        errPipe.fileHandleForReading.readabilityHandler = nil

        stderrLock.lock()
        let stderr = String(data: stderrData, encoding: .utf8) ?? ""
        stderrLock.unlock()
        return RunResult(
            status: process.terminationStatus,
            stdout: String(data: stdoutData, encoding: .utf8) ?? "",
            stderr: stderr
        )
    }
}

File: windows/src/main/services/whisperService.ts
import { app } from 'electron';
import { spawn } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import * as https from 'https';
import { findExecutable } from './toolPaths';

/**
 * Lokale Transkription über whisper.cpp (whisper-cli), analog zur
 * macOS-Version: ffmpeg extrahiert das In/Out-Segment als 16-kHz-WAV,
 * whisper-cli transkribiert nach JSON. Optionales Add-on — ohne
 * installiertes whisper-cli/ffmpeg bleibt die App ansonsten unverändert.
 */

const MODEL_NAME = 'ggml-large-v3-turbo.bin';
const MODEL_URL =
  'https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-large-v3-turbo.bin';

function modelPath(): string {
  return path.join(app.getPath('userData'), 'models', MODEL_NAME);
}

export interface WhisperStatus {
  whisperInstalled: boolean;
  ffmpegInstalled: boolean;
  modelExists: boolean;
  setupProblem: string | null;
}

export function getStatus(): WhisperStatus {
  const whisperInstalled = findExecutable('whisper-cli') !== null;
  const ffmpegInstalled = findExecutable('ffmpeg') !== null;
  const modelExists = fs.existsSync(modelPath());

  let setupProblem: string | null = null;
  if (!whisperInstalled) {
    setupProblem =
      'whisper-cli wurde nicht gefunden. Download der Windows-Release: ' +
      'https://github.com/ggml-org/whisper.cpp/releases (whisper-cli.exe in den PATH legen).';
  } else if (!ffmpegInstalled) {
    setupProblem = 'ffmpeg wurde nicht gefunden. Installation: winget install Gyan.FFmpeg';
  } else if (!modelExists) {
    setupProblem = 'Das Whisper-Modell (~1,6 GB) muss einmalig geladen werden.';
  }
  return { whisperInstalled, ffmpegInstalled, modelExists, setupProblem };
}

export function downloadModel(onProgress: (fraction: number) => void): Promise<void> {
  return new Promise((resolve, reject) => {
    const dest = modelPath();
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    const tmpDest = dest + '.download';
    const file = fs.createWriteStream(tmpDest);

    const request = (url: string) => {
      https
        .get(url, (response) => {
          if (
            response.statusCode &&
            response.statusCode >= 300 &&
            response.statusCode < 400 &&
            response.headers.location
          ) {
            request(response.headers.location);
            return;
          }
          if (response.statusCode !== 200) {
            reject(new Error(`Download fehlgeschlagen (HTTP ${response.statusCode}).`));
            return;
          }
          const total = parseInt(response.headers['content-length'] || '0', 10);
          let received = 0;
          response.on('data', (chunk: Buffer) => {
            received += chunk.length;
            if (total > 0) onProgress(received / total);
          });
          response.pipe(file);
          file.on('finish', () => {
            file.close(() => {
              fs.renameSync(tmpDest, dest);
              resolve();
            });
          });
        })
        .on('error', (err) => {
          fs.unlink(tmpDest, () => {});
          reject(err);
        });
    };
    request(MODEL_URL);
  });
}

export interface WhisperSegment {
  start: number; // Sekunden, relativ zum Video
  text: string;
}

export function transcribe(
  videoPath: string,
  start: number,
  end: number,
  onProgress: (text: string) => void
): Promise<WhisperSegment[]> {
  return new Promise((resolve, reject) => {
    const ffmpeg = findExecutable('ffmpeg');
    const whisper = findExecutable('whisper-cli');
    if (!ffmpeg || !whisper) {
      reject(new Error('whisper-cli oder ffmpeg nicht gefunden.'));
      return;
    }

    const tmpDir = fs.mkdtempSync(path.join(app.getPath('temp'), 'sighting-whisper-'));
    const wavPath = path.join(tmpDir, 'segment.wav');
    const outBase = path.join(tmpDir, 'transcript');

    const cleanup = () => {
      fs.rm(tmpDir, { recursive: true, force: true }, () => {});
    };

    onProgress('Extrahiere Audio …');
    const ff = spawn(ffmpeg, [
      '-hide_banner',
      '-loglevel',
      'error',
      '-nostdin',
      '-ss',
      String(start),
      '-to',
      String(end),
      '-i',
      videoPath,
      '-vn',
      '-ac',
      '1',
      '-ar',
      '16000',
      '-c:a',
      'pcm_s16le',
      wavPath,
    ]);
    let ffError = '';
    ff.stderr.on('data', (d) => (ffError += d.toString()));
    ff.on('close', (code) => {
      if (code !== 0) {
        cleanup();
        reject(new Error(`ffmpeg-Fehler: ${ffError.slice(-400)}`));
        return;
      }

      onProgress('Transkribiere … 0 %');
      const wp = spawn(whisper, [
        '-m',
        modelPath(),
        '-f',
        wavPath,
        '-l',
        'auto',
        '-oj',
        '-of',
        outBase,
        '--print-progress',
      ]);
      let wpError = '';
      wp.stderr.on('data', (d) => {
        const text = d.toString();
        wpError += text;
        const match = text.match(/progress\s*=\s*(\d+)%/);
        if (match) onProgress(`Transkribiere … ${match[1]} %`);
      });
      wp.on('close', (wcode) => {
        if (wcode !== 0) {
          cleanup();
          reject(new Error(`Whisper-Fehler: ${wpError.slice(-400)}`));
          return;
        }
        try {
          const json = JSON.parse(fs.readFileSync(outBase + '.json', 'utf-8'));
          const entries: any[] = json.transcription || [];
          const segments: WhisperSegment[] = entries
            .map((entry) => {
              const text = (entry.text || '').trim();
              const fromMs = entry.offsets?.from;
              if (!text || typeof fromMs !== 'number') return null;
              return { start: start + fromMs / 1000, text };
            })
            .filter((s): s is WhisperSegment => s !== null);
          cleanup();
          resolve(segments);
        } catch (e) {
          cleanup();
          reject(new Error('Whisper-Ausgabe konnte nicht gelesen werden.'));
        }
      });
    });
  });
}
